airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/component/TableSchemaEvolutionClient.kt

```diff
@@ -57,5 +57,156 @@ interface TableSchemaEvolutionClient {
         stream: DestinationStream,
         tableName: TableName,
         columnNameMapping: ColumnNameMapping,
-    )
+    ) {
+        val (actualSchema, actualAdditionalInfo) = discoverSchema(tableName)
+        val (expectedSchema, expectedAdditionalInfo) = computeSchema(stream, columnNameMapping)
+        val schemaDiff = actualSchema.diff(expectedSchema)
+        val additionalInfoDiff = diff(actualAdditionalInfo, expectedAdditionalInfo)
+        applySchemaDiff(
+            stream,
+            columnNameMapping,
+            tableName,
+            expectedSchema,
+            expectedAdditionalInfo,
+            schemaDiff,
+            additionalInfoDiff
+        )
```

Comment:
iceberg has some weirdness, but it follows roughly the same structure. The main difference is that we'd need to change how this piece works: instead of calculating and then immediately applying the diff, iceberg sometimes needs to calculate the diff at the start of the sync, but defer applying it until the end of the sync

also, iceberg's diff tries to compute supertypes, rather than altering directly to the new type - I think that's fine, but probably indicates that `applyDiff` needs to take `actualSchema` as a param? B/c then we can do something like `alterColumn(columnName, computeSupertype(actualSchema[columnName].type, diff[columnName].newType))`

(iceberg, like bigquery, also needs the additionalInfo field - to track the table's "identifier columns", i.e. primary keys)

Reply:
We could break out the underlying methods (discover, diff, etc) to a separate interface if we want to share those but express a different composition for data lakes.

Reply:
yeah, I'm not worried about it. Mostly a question of whether it's worth doing that right now - IMO probably not, since I think we're focusing on clickhouse+snowflake as the main testbeds?